### PR TITLE
docker: update some dockerfiles

### PR DESCRIPTION
**Pull request by asarhaddon**
The modified implementations have been tested on Debian bullseye.

* prefer latest Ubuntu image, hopefully reducing the image count

* for Ubuntu versions, prefer numerical versions to unsorted codenames

* prefer packaged build systems to local rebuilds

* prefer the current default version of each build system

* delegate resolution of indirect dependencies to apt

### CODE REVIEW

File: impls/ocaml/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-[FROM ubuntu:vivid
+[FROM ubuntu:20.04
```
[the remark this answers is not in the document]
Comment (asarhaddon) [reply impls/ocaml/Dockerfile]: It was apparently inoffensive.

[the remark this answers is not in the document]
Comment (asarhaddon) [reply impls/ocaml/Dockerfile]: I am testing manual installations on a physical machine. This does not catch all issues.

**Comment by asarhaddon** [comment]
The xslt failures illustrate such differences.  I have hopefully solved the first issue (python3 may not be in PATH), but now runtest.py fails to find impl/xslt/run on actual docker images.  `make test^xslt` works on my machine. Any idea?

File: impls/ocaml/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (asarhaddon) [reply impls/ocaml/Dockerfile]: From your comment and the error message with the factor language, I guess that Dockerfiles are not executed before each test run, but only used when manually generating a new image for each language.
If this is correct, the results are probably not meaningful, but the merge request at least documents a way to recreate a local test environment, and may be useful during next update of the image.

**Comment by asarhaddon** [comment]
Hello.
While working on #592, I have modified the logo implementation. It now passes tests for step5 with the default garbage collection settings, at least on my machine.
The choice to download the source from Debian instead of upstream reflects my own habits (GPG signature for the downloaded contents, daily tests ensuring that the current version builds with latest toolchains, and so on). I have of course no strong opinion about that.
The Debian package is usually based on the contents of the upstream tarball, including ./configure. I could have avoided the dependency on autoconf, but the other Dockerfiles gave me the impression that  keeping the images minimal was not a priority.

**Comment by dubek** [comment]
Sure.

**Comment by kanaka** [comment]
@asarhaddon BTW, I'm working on getting this refactored on this personal branch: https://github.com/kanaka/mal/tree/update-some-dockerfiles (so that we don't duplicate work). I think I just need to track down a go build issue and then I'll merge since it will be no worse (i.e. elm, common-lisp, and objpascal that have issues).

**Comment by kanaka** [comment]
I have refactored these and merged them to master now. Here is the CI workflow: https://github.com/kanaka/mal/actions/runs/10256887753 Should just be objpascal, elm, and common-lisp that fail.

**Comment by asarhaddon** [comment]
In order to fix objpascal, just remove the `CMem` module from the list of imports at the beginning of each step file.
The common-lisp failure changes when the Dockerfile installs the `cl-ppcre` Debian package but other errors happen before the tests begin. By the way, there also exists a package named cl-quicklisp, but the user must install manually (`https://salsa.debian.org/common-lisp-team/quicklisp/-/blob/master/debian/README.Debian?ref_type=heads`).

**Comment by kanaka** [comment]
@asarhaddon Removing the CMem module did the trick. Out of curiosity, where did you find the info about CMem?

**Comment by kanaka** [comment]
@asarhaddon The common-lisp issue is weird. If I go into the container and run make 5 times, each time it complains about a different dependency and then eventually succeeds. So something isn't quite hooked up right to work with more recent versions of stuff.

**Comment by asarhaddon** [comment]
I have found https://wiki.freepascal.org/CMem while investigating the differences between step0 (not affected) and step1 (affected).  Now, CMem is present in both, and motivated by the C pointers in readline, so trying to remove it was a short in the dark.

**Comment by kanaka** [comment]
@asarhaddon sigh, I got common-lisp working on my local system by using sbcl to build instead of cl-launch: https://github.com/kanaka/mal/commit/062fb6f66ebed857c1d7c3636410557803f35fbf.

However, in GHA CI, it crashes right after building the first step: https://github.com/kanaka/mal/actions/runs/10270914448/job/28419731317

I've filed a ticket and pinged the original author: https://github.com/kanaka/mal/issues/660

**Pull request closed without merging**